SofaKernel/modules/SofaHelper/src/sofa/helper/NameDecoder.h

```diff
@@ -0,0 +1,267 @@
+/******************************************************************************
+*       SOFA, Simulation Open-Framework Architecture, development version     *
+*                (c) 2006-2019 INRIA, USTL, UJF, CNRS, MGH                    *
+*                                                                             *
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+/******************************************************************************
+ * Contributors:
+ *      - damien.marchal@univ-lille.fr
+ *      - jean-nicolas.brunet@inria.fr
+ *****************************************************************************/
+#pragma once
+
+#include <string>
+#include <typeinfo>
+#include <tuple>
+#include <sofa/helper/helper.h>
+namespace sofa::helper
+{
+
+template<class T>
+class HasGetCustomTemplateName
```

Comment:
I guess you didn't like the detector idiom ;-) 

These old c++11 classes could easily be replaced by :

```
    template <class T> 
    using get_custom_template_name_t = decltype(T::Name());
    template <class T> constexpr bool 
    HasGetCustomTemplateName = is_detected_convertible_v<std::string, get_custom_template_name_t, T>;
```
and later on
```
if constexpr (HasGetCustomClassName<T>) {
    (...)
}
```

which also have the advantage of checking that the return type of the method is convertible to a std::string.

Reply:
I prefer much more version based on is_detected_convertible...but it rely on experimental feature isn't ?

Reply:
Awesome, let me know if I can do something to help